core/src/variable.rs

```diff
@@ -84,3 +84,108 @@ impl<T: FromBytes + ToBytes + CLTyped> Instance for Variable<T> {
         namespace.into()
     }
 }
+
+#[cfg(all(feature = "mock-vm", test))]
+mod tests {
+    use crate::{Instance, Variable};
+    use odra_mock_vm::TestEnv;
+    use odra_types::{
+        arithmetic::ArithmeticsError,
+        bytesrepr::{FromBytes, ToBytes},
+        CLTyped, ExecutionError,
+    };
+
+    #[test]
+    fn test_get() {
+        // Given uninitialized var
+        let value = 100;
+        let var = Variable::<u8>::default();
+
+        // When set a value
+        var.set(value);
+
+        // The the value can be returned
+        assert_eq!(var.get().unwrap(), value);
+
+        // When override
+        let value = 200;
+        var.set(value);
+
+        // Then the value is updated
+        assert_eq!(var.get().unwrap(), value);
+    }
+
+    #[test]
+    fn get_default_value() {
+        // Given uninitialized var
+        let var = Variable::<u8>::default();
+
+        // Raw get returns None
+        assert_eq!(var.get(), None);
+        // get_or_default returns the default value
+        assert_eq!(var.get_or_default(), 0);
+    }
+
+    #[test]
+    fn test_add() {
+        // Given var = u8::MAX-1;
+        let initial_value = u8::MAX - 1;
+        let var = Variable::<u8>::init(initial_value);
+
+        // When add 1
+        var.add(1);
+
+        // Then the value should be u8::MAX
+        assert_eq!(var.get_or_default(), initial_value + 1);
+
+        // When add 1 to max value
+        // Then should revert
+        TestEnv::assert_exception(
+            Into::<ExecutionError>::into(ArithmeticsError::AdditionOverflow),
+            || {
+                var.add(1);
+            },
+        );
+    }
+
+    #[test]
+    fn test_subtract() {
+        // Given var = 2;
+        let initial_value = 2;
+        let var = Variable::<u8>::init(initial_value);
+        // When subtract 1
+        var.subtract(1);
+
+        // Then the value should be reduced by 1
+        assert_eq!(var.get_or_default(), initial_value - 1);
+
+        // When subtraction causes overflow
+        // Then it reverts
+        TestEnv::assert_exception(
+            Into::<ExecutionError>::into(ArithmeticsError::SubtractingOverflow),
+            || {
+                var.subtract(2);
+            },
+        );
+    }
+
+    impl<T> Default for Variable<T>
```

Comment:
I think those implementations should be a simple `setup` function. It introduces unncessary complexity.

Reply:
But having the default implementation the code is more "rusty" and more verbose IMO. If I write `let var = Variable::<u8>::default()` it's clear I'm creating an instance of variable, a mysterious function `setup()` wouldn't give that feeling. :) If I were doing some more complex setup, then fine I would name it `setup()`.

Reply:
By implementing `Default` in tests, you do it for every test code in the crate. It means eg in `mapping.rs` I could write:
```rust
let v = Variable::<u8>::default()
```
And it will work. I just want to be sure you understand that you introduce default behaviours for `Variable` for all test code.

Another problem I see is that it's not clear that in this code:
```rust 
let v = Variable::<u8>::default()
let x = Variable::<u8>::default()
```
`v` and `x` points at the same storage. Could be confusing.

Reply:
I would leave this notatnion:
```rust
let x = Variable::<u8>::instance("x");
let y = Variable::<u8>::instance("something else");
```

Reply:
The fact you could doesn't mean you should and why would you?
I can create an instance of Variable effortlessly.

But I must admit, I wasn't thinking I was creating an implementation that is available across the crate.

You could copy-paste and write.
```
let v = Variable::<u8>::instance("a");
let x = Variable::<u8>::instance("a");
```
The code compiles, but obviously, it wasn't your intention.

Reply:
Actaully this case should also be tested.
```rust
let v = Variable::<u8>::instance("a");
let x = Variable::<u8>::instance("a");
```
We will use it as a feature when we will introduce using the same module inside two other modules. Eg: two modules could share the same AccessControl module.

Reply:
You know, the main reason was I didn't want to write `let x = Variable::<u8>::instance("x");` because this `x` literal doesn't matter, I could pass even an empty string and it would (probably) work. :)

Reply:
> Actaully this case should also be tested.
> 
> ```rust
> let v = Variable::<u8>::instance("a");
> let x = Variable::<u8>::instance("a");
> ```
> 
> We will use it as a feature when we will introduce using the same module inside two other modules. Eg: two modules could share the same AccessControl module.

Ok, here we agree. :grinning:

Reply:
PR is approved. Your call now.